imap/imapd.c

```diff
@@ -6114,21 +6191,121 @@ static void cmd_search(char *tag, int usinguid)
+             * implies the "ALL" result option.
+             */
+            searchargs->returnopts = SEARCH_RETURN_ALL;
         }
-        else {
-            prot_printf(imapd_out, "%s NO %s\r\n", tag, error_message(r));
+
+        /* Cycle through each of the possible source filters */
+        for (mrock.filter = SEARCH_SOURCE_SELECTED;
+             mrock.filter <= SEARCH_SOURCE_MAILBOXES; mrock.filter <<= 1) {
+
+            if (!(searchargs->filter & mrock.filter)) continue;
+
+            switch (mrock.filter) {
+            case SEARCH_SOURCE_SELECTED:
+                if (!imapd_index) {
+                    prot_printf(imapd_out,
+                                "%s BAD Please select a mailbox first\r\n", tag);
+                    goto done;
+                }
+
+                if (!index_check(imapd_index, 0, 0)) {
+                    n += index_search(imapd_index, searchargs, /* usinguid */1);
+
+                    hash_insert(index_mboxname(imapd_index),
+                                (void *) 1, &mrock.mailboxes);
+                }
+                break;
+
+            case SEARCH_SOURCE_PERSONAL:
+            case SEARCH_SOURCE_INBOXES:
+                mboxlist_usermboxtree(searchargs->userid, searchargs->authstate,
+                                      multisearch_cb, &mrock, 0);
+                break;
+
+            case SEARCH_SOURCE_SUBSCRIBED:
+                mboxlist_usersubs(searchargs->userid, multisearch_cb, &mrock, 0);
+                break;
+
+            default: {
+                strarray_t *mailboxes = NULL;
+                unsigned flags = 0;
+                int i;
+
+                switch (mrock.filter) {
+                case SEARCH_SOURCE_SUBTREE:
+                    mailboxes = &searchargs->subtree;
+                    break;
+
+                case SEARCH_SOURCE_SUBTREE_ONE:
+                    mailboxes = &searchargs->subtree_one;
+                    break;
+
+                case SEARCH_SOURCE_MAILBOXES:
+                    /* Just the mailbox - no children */
+                    flags = MBOXTREE_SKIP_CHILDREN;
+                    mailboxes = &searchargs->mailboxes;
+                    break;
+                }
+                
+                /* Cycle through each mailbox [tree] */
+                for (i = 0; i < strarray_size(mailboxes); i++) {
+                    const char *intname = strarray_nth(mailboxes, i);
+
+                    if (mrock.filter == SEARCH_SOURCE_SUBTREE_ONE) {
+                        /* Calculate the depth of the "root" mailbox */
+                        mbname_t *mbname = mbname_from_intname(intname);
+                        mrock.root_depth = strarray_size(mbname_boxes(mbname));
+                        mbname_free(&mbname);
+                    }
+
+                    mboxlist_mboxtree(intname, multisearch_cb, &mrock, flags);
+                }
+                break;
+            }
+            }
```

Comment:
nit: indenting

Reply:
@rsto This is an artifact of declaring variables inside a switch clause.  Its looks weird, but this is how Emacs does the indenting.  How would you like this to look>

Reply:
I don't like the two `}`s at the same indentation level either, but I thought about this myself, and concluded that it's an artefact of our style placing the `case` labels at the same indent level as the `switch` itself.

If our `switch` style was like:

```
switch(foo) {
    case 'a':
        // something simple;
        break;
    case 'b': {
        // something complex
        break;
    }
}
```

-- that is, with the cases at their own indent level -- then this wouldn't be a problem.  But that would use up a lot of horizontal space on whitespace, leaving less for code, meaning more lines need to be wrapped, leading to more complicated wrapping solutions...

I suppose the variable(s) in question could be moved to the containing scope, removing the need for `{}` in the `case`, but as it is their lifetime is clear, even if the `}}` is jarring.  If they were up a level, one would then need to read that whole scope holding them in mind, even though they're only used briefly in that one place.

So while I don't really like it, I didn't have a clearly better suggestion, and still don't.